lib/vm-agent.js

```diff
@@ -9,450 +9,696 @@
+        callback();
+        return;
+    }
 
-/*
- * Initializes the serial queue for a single uuid and pushes the item to be
- * processed.
- */
-VmAgent.prototype.pushSample = function (options) {
-    var uuid = options.uuid || this.uuid;
+    // On any error when we don't already have a timer running, we set a
+    // timer to re-queue the VM. This is always safe since in the worst
+    // case we do an extra update with the latest data.
+    if (!self.retryDelays[vm_uuid]) {
+        self.retryDelays[vm_uuid] = {delay: INITIAL_UPDATE_DELAY};
+    }
+    delay = self.retryDelays[vm_uuid].delay;
 
-    if (this.queues[uuid] === undefined) {
-        this.queues[uuid] = this.createQueue(uuid);
+    // Increment for next time.
+    self.retryDelays[vm_uuid].delay *= 2;
+    if (self.retryDelays[vm_uuid].delay > MAX_UPDATE_DELAY) {
+        self.retryDelays[vm_uuid].delay = MAX_UPDATE_DELAY;
     }
 
-    this.queues[uuid].push(options);
-};
+    self.log.warn({err: err, vm_uuid: vm_uuid}, 'update failed, will try '
+        + 'again in ' + delay + ' ms');
+
+    self.retryDelays[vm_uuid].timer = setTimeout(function _delayedRetry() {
+        if (self.retryDelays[vm_uuid]) {
+            delete self.retryDelays[vm_uuid].timer;
+        }
+        self.queueVm(vm_uuid);
+    }, delay);
 
+    callback();
+};
 
 /*
- * Retries an update operation.
+ * The fact that we use this queue serially has the additional advantage
+ * currently of debouncing updates that are coming in frequently. For example if
+ * a VM is under heavy modification and posting 'modify' events multiple times
+ * per second, most of the time we'll never have more than one update pending in
+ * the queue.
  */
-VmAgent.prototype.retryUpdate = function (fn, options, callback) {
-    var log = this.log;
-    var retryOpts = { initialDelay: 2000, maxDelay: 64000 };
-
-    function logAttempt(aLog, host) {
-        function _log(number, delay, err) {
-            var level;
-            if (number === 0) {
-                level = 'info';
-            } else if (number < 5) {
-                level = 'warn';
-            } else {
-                level = 'error';
-            }
-
-            aLog.error(err, 'retry error for %s', options.uuid);
-            aLog[level]({
-                ip: host,
-                attempt: number,
-                delay: delay
-            }, 'retry attempt for %s', options.uuid);
+VmAgent.prototype.queueVm = function queueVm(vm_uuid) {
+    var self = this;
+    var alreadyQueued = false;
+    var queueIdx;
+
+    // If we don't already have a queued update for this VM, queue one.
+    //
+    // NOTE: the queue.queued does not include any vm_uuid that is currently
+    // being processed, which is fine because we may be past the lookup and
+    // in the waiting for VMAPI phase. We need to update again when we can.
+    //
+    for (queueIdx in self.queue.queued) {
+        if (self.queue.queued[queueIdx].task === vm_uuid) {
```

Comment:
The README does list them as "Read-only public properties (for debugging)" so I assumed the public there meant they were part of the API, and yeah I didn't want to have to keep another set of data in sync with that when it was already there.